[TSAN][live speech translation] Fix A data race in caffe2

Polyomino opened the pull request:
Summary: noticed that context quantized_engine is accessed and written from multiple threads

Test Plan:
➜  fbsource buck test --flagfile fbcode/mode/dev-tsan //xplat/assistant/integration_test/tests/supernova/speechtranslation:live_speech_translation_en_fr_tests -- --exact 'fbsource//xplat/assistant/integration_test/tests/supernova/speechtranslation:live_speech_translation_en_fr_tests - Translate/LiveSpeechTranslationTests.LiveSpeechTranslationEnFr/silence___fr_en'

Rollback Plan:

Differential Revision: D76921416

Polyomino commented:
/easycla

Polyomino commented on aten/src/ATen/Context.cpp:
```diff
@@ -535,13 +536,14 @@ at::QEngine Context::qEngine() const {
   }();
-  return quantized_engine.value_or(_quantized_engine);
+  auto qt_engine = quantized_engine.load();
+  return qt_engine == at::kNoQEngine ? _quantized_engine : qt_engine;
```
I think the code before would pick quantized_engine if it's available, then fall back to _quantized engine. 
So in the new code, If quantized_engine is at::kNoQEngine we pick _quantized_engine otherwise quantized_engine

jerryzh168 replied:
oh OK, makes sense, since nullopt becomes kNoQEngine now

jerryzh168 commented:
can you fix failed tests?

Polyomino commented:
@pytorchbot label "topic: not user facing"

Polyomino commented:
@jerryzh168 how reliable are these perf benchmarks? Seems like more instruction overhead than I'd expect.

https://github.com/pytorch/pytorch/actions/runs/15788568689/job/44513133975?pr=156378

jerryzh168 commented:
> @jerryzh168 how reliable are these perf benchmarks? Seems like more instruction overhead than I'd expect.
> 
> [pytorch/pytorch/actions/runs/15788568689/job/44513133975?pr=156378](https://github.com/pytorch/pytorch/actions/runs/15788568689/job/44513133975?pr=156378)

I'm not aware that we are tracking performance before

Polyomino commented:
@jerryzh168 does this need approval again?

Polyomino committed:
[TSAN][live speech translation] Fix A data race in caffe2 (#156378)

Summary:
noticed that context quantized_engine is accessed and written from multiple threads

Pull Request resolved: https://github.com/pytorch/pytorch/pull/156378

Test Plan:
➜  fbsource buck test --flagfile fbcode/mode/dev-tsan //xplat/assistant/integration_test/tests/supernova/speechtranslation:live_speech_translation_en_fr_tests -- --exact 'fbsource//xplat/assistant/integration_test/tests/supernova/speechtranslation:live_speech_translation_en_fr_tests - Translate/LiveSpeechTranslationTests.LiveSpeechTranslationEnFr/silence___fr_en'

Rollback Plan:

Rollback Plan:

Reviewed By: jerryzh168

Differential Revision: D76921416

Pulled By: Polyomino

cyyever commented:
@pytorchbot merge -i
